### PR TITLE
Do not unnecessarily connect to the privacyIDEA server

### PR DESCRIPTION
**Pull request by gber**

### CODE REVIEW

File: personal/privacyidea/class_PrivacyIdeaUtils.inc [hunk 1]
```diff
@@ -136,6 +134,10 @@ class PrivacyIdeaUtils implements PILog
      */
     public function countUserTokens($tokens)
     {
+        if (!$this->authenticate()) {
+            return;
+        }
```
Comment (sunweaver) [inline personal/privacyidea/class_PrivacyIdeaUtils.inc:139]: What will happen if we return from here. Is there already some auth-failure handling (REST API auth against privacyIDEA failing, I mean).

@dzatoah cc

File: personal/privacyidea/class_PrivacyIdeaUtils.inc [hunk 2]
```diff
@@ -176,6 +178,10 @@ class PrivacyIdeaUtils implements PILog
         assert('string' === gettype($description));
         assert($tokenType == "paper" || $tokenType == "totp" || $tokenType == "webauthn");
 
+        if (!$this->authenticate()) {
+            return array();
+        }
+
```
Comment (sunweaver) [inline personal/privacyidea/class_PrivacyIdeaUtils.inc:184]: dito here

File: personal/privacyidea/class_PrivacyIdeaUtils.inc [hunk 3]
```diff
@@ -224,6 +230,10 @@ class PrivacyIdeaUtils implements PILog
         assert(!empty($tokenCode));
         assert($tokenType == "paper" || $tokenType == "totp");
 
+        if (!$this->authenticate()) {
+            return array();
+        }
+
```
Comment (sunweaver) [inline personal/privacyidea/class_PrivacyIdeaUtils.inc:236]: dito here

**Review by sunweaver — COMMENTED** [review]
Some feedback needed...

File: personal/privacyidea/class_PrivacyIdeaUtils.inc (hunk 1 above)
Comment (gber) [reply personal/privacyidea/class_PrivacyIdeaUtils.inc]: This method is public and there is a TODO to modify this so it fetches tokens directly from PrivacyIDEA without passing them in. However it is currently only used internally and needs to be passed tokens in, so the check can be left out for now until that is actually implemented.

File: personal/privacyidea/class_PrivacyIdeaUtils.inc (hunk 2 above)
Comment (gber) [reply personal/privacyidea/class_PrivacyIdeaUtils.inc:180]: This is already an existing  failure mode. However while reviewing this again I found a bug in `personal/privacyidea/tokens/class_MFAPaperToken.inc` where an error wasn't handled properly. I've fixed that in another commit.

File: personal/privacyidea/class_PrivacyIdeaUtils.inc (hunk 3 above)
Comment (gber) [reply personal/privacyidea/class_PrivacyIdeaUtils.inc:232]: The above is equal to the existing failure mode of `piSendRequest` and handled properly.

**Comment by gber** [comment]
While addressing the review I found another bug in `personal/privacyidea/class_mfaAccount.inc` where failure in `getTokensOfUser` wasn't handled properly, a fix is also included here ([61236a1](https://github.com/gosa-project/gosa-plugins-privacyidea/pull/10/commits/61236a1afe6569b4bb7a6e8a4400352e4b2c14e3)).

**Pull request merged**